weldx/asdf/validators.py

```diff
@@ -95,6 +95,261 @@ def _unit_validator(
+    _int:
+        Integer
+    exp_string:
+        String with the expected dimension
+
+    Returns
+    -------
+    bool
+        True or False
+    """
+    if _int < 0:
+        raise ValueError("Negative dimension found")
+    if ":" in exp_string:
+        ranges = exp_string.split(":")
+        if ranges[0] == "" and ranges[1] == "":
+            return True
+        else:
+            if ranges[0] != "" and ranges[1] != "":
+                if int(ranges[0]) <= int(ranges[1]):
+                    pass
+                else:
+                    raise ValueError(
+                        f"The range should not be descending in {exp_string}"
+                    )
+            if ranges[0] != "":
+                return 0 <= int(ranges[0]) <= _int
+            if ranges[1] != "":
+                return int(ranges[1]) >= _int > 0
+    else:
+        return _int == int(exp_string)
+
+
+def _prepare_list(_list, list_expected):
+    """Prepare a List and an expected List for validation.
+
+    parameters
+    ----------
+    _list:
+        List with values
+    list_expected:
+        List with expected values
+    returns
+    -------
+    _list:
+        prepared List
+    list_expected:
+        prepared List with expected values
+    """
+    # remove blank spaces in dict_test
+    _list = [x.replace(" ", "") if isinstance(x, str) else x for x in _list]
+    # accept "~" additionally as input of ":". And remove blank spaces.
+    list_expected = [
+        x.replace(" ", "").replace("~", ":") if isinstance(x, str) else x
+        for x in list_expected
+    ]
+    # turn around the list if "..." or "(" are at the beginning.
+    # because the validation is made from begin -> end.
+    # like this we validate the array from end -> begin.
+    if "(" in str(list_expected[0]) or "..." in str(list_expected[0]):
+        list_expected = list(reversed(list_expected))
+        _list = list(reversed(_list))
+
+    return _list, list_expected
+
+
+def _validate_expected_list(list_expected):
+    """Validate an expected List and raises exceptions.
+
+
+    params
+    ------
+    list_expected:
+        Expected List to validate against
+
+    raises
+    ------
+    ValueError:
+        ValueError will be raised if an rule violation is found
+    """
+    validator = 0
+    for exp in list_expected:
+        if validator == 1:
+            if "(" not in str(exp):
+                raise ValueError(
+                    "Optional dimensions in the expected "
+                    "shape should only stand at the end/beginning."
+                )
+        elif validator == 2:
+            raise ValueError('After "..." should not be another dimension.')
+        elif "..." in str(exp):
+            if "..." != exp:
+                raise ValueError(
+                    f'"..." should not have additional properties:' f" {exp} was found."
+                )
+            validator = 2
+        elif "(" in str(exp):
+            validator = 1
```

Comment:
consider merging the first 2 `if` statements. Then you can replace the `elif`s with `if`s because a `raise` will always prevent further code execution. Then also merge these 2 statements 

~~~ python
elif "..." in str(exp):
            if "..." != exp:
~~~

into a single `if` statement.